nixops/resources/vpc.py

```diff
@@ -0,0 +1,275 @@
+            after=[self.handle_create],
+            handle=self.realize_associate_ipv6_cidr_block)
+        self.handle_tag_update = Handler(['tags'], after=[self.handle_create], handle=self.realize_update_tag)
+
+    def show_type(self):
+        s = super(VPCState, self).show_type()
+        region = self._state.get('region', None)
+        if region: s = "{0} [{1}]".format(s, region)
+        return s
+
+    @property
+    def resource_id(self):
+        return self._state.get('vpcId', None)
+
+    def prefix_definition(self, attr):
+        return {('resources', 'vpc'): attr}
+
+    def get_physical_spec(self):
+        return { 'vpcId': self._state.get('vpcId', None)}
+
+    def get_definition_prefix(self):
+        return "resources.vpc."
+
+    def connect(self):
+        if self._client: return
+        assert self._state['region']
+        (access_key_id, secret_access_key) = nixops.ec2_utils.fetch_aws_secret_key(self.access_key_id)
+        self._client = boto3.session.Session().client('ec2', region_name=self._state['region'], aws_access_key_id=access_key_id, aws_secret_access_key=secret_access_key)
+
+    def _destroy(self):
+        if self.state != self.UP: return
+        self.connect()
+        self.log("destroying vpc {0}...".format(self._state['vpcId']))
+        try:
+            self._client.delete_vpc(VpcId=self._state['vpcId'])
+        except botocore.exceptions.ClientError as e:
+            if e.response['Error']['Code'] == 'InvalidVpcID.NotFound':
+                self.warn("vpc {0} was already deleted".format(self._state['vpcId']))
+            else:
+                raise e
+
+        with self.depl._db:
+            self.state = self.MISSING
+            self._state['vpcId'] = None
+            self._state['region'] = None
+            self._state['cidrBlock'] = None
+            self._state['instanceTenancy'] = None
+            self._state['enableDnsSupport'] = None
+            self._state['enableDnsHostname'] = None
+            self._state['enableVpcClassicLink'] = None
+
+    def create_after(self, resources, defn):
+        return {r for r in resources if
+                isinstance(r, nixops.resources.elastic_ip.ElasticIPState)}
+
+    def create(self, defn, check, allow_reboot, allow_recreate):
+        diff_engine = self.setup_diff_engine(config=defn.config)
+
+        self.access_key_id = defn.config['accessKeyId'] or nixops.ec2_utils.get_access_key_id()
+        if not self.access_key_id:
+            raise Exception("please set 'accessKeyId', $EC2_ACCESS_KEY or $AWS_ACCESS_KEY_ID")
+
+        for handler in diff_engine.plan():
+            handler.handle(allow_recreate)
+
+        self.ensure_state_up(check)
+
+    def ensure_state_up(self, check):
+        config = self.get_defn()
+        self._state["region"] = config['region']
+        self.connect()
+        # handle vpcs that are deleted from outside nixops e.g console
+        if self._state.get('vpcId', None):
+            if check:
```

Comment:
I wonder if should get rid of the `check` parameter, given that nowadays we have a separate `MachineState._check()` method. That is, `nixops deploy --check` could be implemented by first calling `_check()` on each resource to verify each resource and update its state to match its actual condition, and then calling `create` as usual (but without a `check` flag).

There is a FIXME in `MachineState._check()` that it should be moved to `ResourceState` BTW.

Reply:
@edolstra Couldn't agree more. I was even thinking at some point that we should add `nixops refresh` as a separate command which takes care of updating the state. Didn't know about _check method though, so will definitely use it.